### PR TITLE
Add handling more than 100 pods/nodes to Core Metrics API

### PR DESCRIPTION
**Pull request by znirzej**
Feature & Tests - starting with synchronous implementation

### CODE REVIEW

**Comment by serathius** [comment]
Would be great to introduce a test for this

File: custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client.go [hunk 1]
```diff
@@ -27,53 +28,100 @@ func newClient(translator *translator.Translator) *stackdriverCoreClient {
 	return &stackdriverCoreClient{translator}
 }
 
-// TODO(holubowicz): handle a case when len(resourceNames) > oneOfMax
-func (p *stackdriverCoreClient) getContainerCPU(podsNames []string) (map[string]map[string]resource.Quantity, map[string]api.TimeInfo, error) {
-	stackdriverRequest, err := p.translator.GetSDReqForContainersWithNames(podsNames, containerCPUMetricName, metricKindCPU, labels.Everything(), translator.AllNamespaces)
-	if err != nil {
-		return nil, nil, err
+func (p *stackdriverCoreClient) getPodMetric(podsNames []string, metricName string, metricKind string, labels labels.Selector) (map[string]map[string]resource.Quantity, map[string]api.TimeInfo, error) {
+	numOfRequests := (len(podsNames) + translator.MaxNumOfArgsInOneOfFilter - 1) / translator.MaxNumOfArgsInOneOfFilter // ceil
```
Comment (serathius) [inline custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client.go:32]: I think we should restrict maximum number of pararel requests. e.g. setup pool of 5 goroutines that read from channel

File: custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client.go [hunk 2]
```diff
@@ -27,53 +28,100 @@ func newClient(translator *translator.Translator) *stackdriverCoreClient {
...
+
+	for i := 0; i < numOfRequests; i++ {
+		segmentBeg := i * translator.MaxNumOfArgsInOneOfFilter
+		segmentEnd := min((i+1)*translator.MaxNumOfArgsInOneOfFilter, len(podsNames))
+		stackdriverRequest, err := p.translator.GetSDReqForContainersWithNames(podsNames[segmentBeg:segmentEnd], metricName, metricKind, labels, translator.AllNamespaces)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		go func(stackdriverRequest *stackdriver.ProjectsTimeSeriesListCall) {
```
Comment (serathius) [inline custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client.go:47]: I think paralyzing requests is a good end goal, but it can make code pretty complicated and hard to spot bugs. Could we start synchronous implementation ?

File: custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client.go [hunk 3]
```diff
@@ -27,53 +28,100 @@ func newClient(translator *translator.Translator) *stackdriverCoreClient {
...
+			errChannel <- err
+		}(stackdriverRequest)
 	}
-	stackdriverResponse, err := stackdriverRequest.Do()
-	if err != nil {
-		return nil, nil, err
+	nodeMetric := make(map[string]resource.Quantity)
+	timeInfo := make(map[string]api.TimeInfo)
+
+	for i := 0; i < numOfRequests; i++ {
```
Comment (serathius) [inline custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client.go:104]: This will wait infinitly if one of the goroutines panics, please use WaitGroup

File: custom-metrics-stackdriver-adapter/pkg/adapter/translator/response_translator.go [hunk 4]
```diff
@@ -125,7 +125,11 @@ func (t *Translator) GetMetricsFromSDDescriptorsResp(response *stackdriver.ListM
 func (t *Translator) GetCoreContainerMetricFromResponse(response *stackdriver.ListTimeSeriesResponse) (map[string]map[string]resource.Quantity, map[string]api.TimeInfo, error) {
 	containerMetric := make(map[string]map[string]resource.Quantity)
 	timeInfo := make(map[string]api.TimeInfo)
+	return t.GetCoreContainerMetricFromResponseAppend(containerMetric, timeInfo, response)
+}
 
+// GetCoreContainerMetricFromResponseAppend for each pod adds to map entries from response about metric and TimeInfo.
+func (t *Translator) GetCoreContainerMetricFromResponseAppend(containerMetric map[string]map[string]resource.Quantity, timeInfo map[string]api.TimeInfo, response *stackdriver.ListTimeSeriesResponse) (map[string]map[string]resource.Quantity, map[string]api.TimeInfo, error) {
```
Comment (serathius) [inline custom-metrics-stackdriver-adapter/pkg/adapter/translator/response_translator.go:132]: Mutating arguments is not the best pattern, could we extract for loop outside?

File: custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client.go [hunk 5]
```diff
@@ -27,53 +28,100 @@ func newClient(translator *translator.Translator) *stackdriverCoreClient {
...
+		stackdriverRequest, err := p.translator.GetSDReqForNodesWithNames(nodeNames[segmentBeg:segmentEnd], metricName, metricKind, labels)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		go func(stackdriverRequest *stackdriver.ProjectsTimeSeriesListCall) {
+			metrics, err := stackdriverRequest.Do()
+			responseChannel <- metrics
+			errChannel <- err
+		}(stackdriverRequest)
```
Comment (serathius) [inline custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client.go:99]: gj, using proper closure

File: custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client.go (hunk 2 above)
Comment (znirzej) [reply custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client.go:47]: Fine, we can start with adding synchronous version

File: custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client_test.go [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+	{
+		numOfNodes: 834,
+	},
+}
+
+func TestCoreprovider_getContainerMetric(t *testing.T) {
+	tr, _ :=
+		translator.NewFakeTranslator(2*time.Minute, time.Minute, "my-project", "my-cluster", "my-zone", time.Date(2017, 1, 2, 13, 2, 0, 0, time.UTC), true)
+
+	for _, testType := range testsTypes {
```
Comment (serathius) [inline custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client_test.go:119]:
```suggestion
	for _, metric := range []string{"CPU", "Memory"} {
```

File: custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client_test.go [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+							entry = getContainerGuageResponseEntry(podId, containerId)
+						} else {
+							entry = getContainerCumulativeResponseEntry(podId, containerId)
+						}
+						possibleResponses = append(possibleResponses, entry)
+						if containerId == 1 {
+							allPods = append(allPods, entry.resourceName)
+						}
+					}
+				}
```
Comment (serathius) [inline custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client_test.go:138]: This looks like a setup creating list of pods, can you extract a function here?

File: custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client_test.go [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+					}
+
+					expectedTime, err := time.Parse(time.RFC3339, entry.response.Points[0].Interval.EndTime)
+					if err != nil {
+						t.Fatalf("Internal Error, can't parse expected time")
+					}
+					if !reflect.DeepEqual(timeResult[key].Timestamp, expectedTime) {
+						t.Errorf("TimeInfo for %v metric - expected: %v, but have %v", *entry.response.Resource, expectedTime, timeResult[key].Timestamp)
+					}
+				}
```
Comment (serathius) [inline custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client_test.go:174]: I don't understand what your testing. 

For code that splits list (all pods) into multiple lists (pods in single query) of size 100 I would expect something like:
* Test if length of requests is correct (all requests beside last one is size 100, and sum matches)
* Test if all pods are tested (throw pod names into map and check if size stays the same)

File: custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client.go [hunk 9]
```diff
@@ -21,59 +22,80 @@ const (
 
 type stackdriverCoreClient struct {
 	translator *translator.Translator
+	doRequest  func(*stackdriver.ProjectsTimeSeriesListCall) (*stackdriver.ListTimeSeriesResponse, error) // for testing proposes
```
Comment (serathius) [inline custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client.go:25]: Please create interface for that.

File: custom-metrics-stackdriver-adapter/pkg/adapter/translator/response_translator_core.go [hunk 10]
```diff
@@ -0,0 +1,128 @@
+package translator
```
Comment (serathius) [inline custom-metrics-stackdriver-adapter/pkg/adapter/translator/response_translator_core.go:1]: Would it make sense to move refactor to separate PR?

File: custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client_test.go (hunk 8 above)
Comment (znirzej) [reply custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client_test.go]: Now I'm testing:
- If every pod was in request exactly one time
- If there is no request with more than 100 pods
- If the number of requests is minimal possible (equal to ceil(#pods / 100))
- If every container in included into result - it may be possible, that every pod was requested as required but it's result wasn't included in response - testing only requests don't cover all code

I think that better than require exact structure of request (100, 100, 100, ..., #pods % 100) It's better to check listed requirements. There could be another implementation which divides pod into equal-size groups and it's also a good implementation.

**Comment by serathius** [comment]
LGTM,
/cc @kawych
